### DISCUSSION TITLE
BadSessionNotActivated issue

### DISCUSSION THREAD

**Discussion by m4rkisgrov3** [FreeOpcUa/python-opcua#1297] [Q&A]
Hello, please can anyone help?
I'm using the python-opcua library in a fairly simple way, so I've been surprised when I encountered this error.

I've written a Python client that subscribes to about 160 nodes. I create 1 handler, and subscribe to all the nodes in one go. As follows:

        self.client = Client(self.opcserverurl)
        self.client.connect()
        validnodes = []
        ...
        _In a for loop I now populate the valid nodes list:_
        for...
                node = self.client.get_node(nodename)
                validnodes.append(node)
        ...
        handler = SubscribeHandler()
        sub = self.client.create_subscription(100, handler)
        sub.subscribe_data_change(validnodes)

Of course, when any of the node values change, the handler is called. This has worked quite happily until I ramp up the number of nodes I subscribe to; I now subscribe to 160, I thought I'd be able to subscribe to 1000s. Also, note the 100 period value in create_subscription.

So, now when I run this whilst subscribing to more nodes, it falls over on the subscription with this error message:

_exception calling callback for <Future at Oxffff98a23bb0 state=finished returned Buffer>
Traceback (most recent call last):
    File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 328, in _invoke_callbacks
        callback (self)
    File "/root/.local/lib/python3.8/site-packages/opcua/client/ua_client.py", line 493, in call publish callback
        self._uasocket _uasocket.check_answer (data, "while waiting for publish response")
    File "/root/.local/lib/python3.5/site-packages/opcua/client/ua_client.py", line 93, in check_answer
        hdr. ServiceResult.check()
    File "/root/.local/lib/python3.5/site-packages/opcua/ua/uatypes.py", line 218, in check
        raise UaStatusCodeError (self.value)
**opcua.ua.uaerrors.auto.BadSessionNotActivated: "The session cannot be used because Activate Session has not been called." (BadsessionNotActivated)**_

Note the error in the traceback, BadSessionNotActivated.

Please can someone help, with less nodes I subscribe to, all this works well and all data changes are captured, when I subscribe to more than 160 nodes, I get this error.

I'm thinking this could be a timing issue? I specify a value of 100 for the period when I create the subscription, is this enough? Or, another thing that crossed my mind is that should I be creating a session since it's telling me BadSessionNotActivated?

Many thanks for help, happy to provide more info if required.

**Comment by AndreasHeine** [comment]
what is the opcua server? 
how many sessions does the server allow?
did you check the servers capabilities? 
how many monitored items does he allow per subscription?
have you checked wireshark trace to get a clue who sends the error (i assume the server)?

**Reply by m4rkisgrov3** [reply]
Many thanks for the reply. Here's some answers I can get right now...

what is the opcua server? _**Siemens S7-1510 CPU w. FW v2.6.**_

how many sessions does the server allow? **_10 (configured), 32 (system maximum)_**

how many monitored items does the server allow per subscription?
Limitations of the OPC UA Server for S7-1500 CPU’s can be found here (Table 1, column 1 in our case)… https://support.industry.siemens.com/cs/document/109755846/what-are-the-system-limits-of-the-opc-ua-server-with-s7-1500-and-s7-1200-?dti=0&lc=en-US

Here's a snippet from the system limits table in the link above:
**Maximum number of subscriptions with more than 1000 monitored elements per subscription over all sessions = 10
Maximum number of subscriptions with a maximum of 1000 monitored elements per subscription over all sessions = 20
Maximum number of subscriptions per session = 20
Smallest possible sampling interval = 100 ms
Smallest possible Send interval = 500 ms
Maximum number of sessions = 32**

Here's my server side config:
![image003](https://user-images.githubusercontent.com/74838433/112122753-80700380-8bb8-11eb-8e99-d08708bba17e.png)

It seems to me that I'm well within the limitations so I don't quite understand why I seeing this error. I'm way below the number of nodes to subscribe to, and I'm only creating one session (in fact, I'm not creating any sessions so I presume 1 session).

Things I'm thinking of trying; create sessions on the client side, limit the number of nodes per subscription, to say 50? Would these be sensible way to tackle this?

As for question around using wireshark, that'll take more digging, but I'm sure this is a server side issue.

Many thanks for help, very much appreciate it.
Mark

**Reply by AndreasHeine** [reply]
could you provide a wireshark trace?

the weired thing is that the opcua client just asks for changes and get all datachanges in one response back so we need to figure out why the session is closed or at least not active!?

**Reply by m4rkisgrov3** [reply]
Thanks Andreas.
I don't have access to the server so I'm going to struggle to get a wireshark trace to determine exactly what's going on. I write the client side software and basically hand it over.
When you talk about sessions, is that a session on the client side or the server side?
Could it be a problem where the client side just cannot cope with all the responses when I ramp it up?
Thanks.

**Reply by AndreasHeine** [reply]
turn on the logger in debug mode on client side and post the output:
    logging.basicConfig(level=logging.WARN)
    logger = logging.getLogger("opcua")
    logger.setLevel(logging.DEBUG)

its hard to determine where the issue is...

**Reply by m4rkisgrov3** [reply]
Yes, thanks Andreas, I'll do that...

**Reply by m4rkisgrov3** [reply]
Hi Andreas, I've added the logging to my OPC client, I how have the error output, as follows:

INFO:opcua.client.ua_client:publish
DEBUG:opcua.client.ua_client.Socket:Sending: PublishRequest(TypeId:{self.TypeId}, RequestHeader:{self.RequestHeader}, Parameters:{self.Parameters})
ERROR:opcua.client.ua_client:Exception while calling user callback: %s
Traceback (most recent call last):
  File "/root/.local/lib/python3.8/site-packages/opcua/client/ua_client.py", line 535, in _call_publish_callback
    callback(response.Parameters)
  File "/root/.local/lib/python3.8/site-packages/opcua/common/subscription.py", line 140, in publish_callback
    self.server.publish([ack])
  File "/root/.local/lib/python3.8/site-packages/opcua/client/ua_client.py", line 482, in publish
    self._uasocket.send_request(request, self._call_publish_callback, timeout=0)
  File "/root/.local/lib/python3.8/site-packages/opcua/client/ua_client.py", line 81, in send_request
    future = self._send_request(request, callback, timeout, message_type)
  File "/root/.local/lib/python3.8/site-packages/opcua/client/ua_client.py", line 72, in _send_request
    self._socket.write(msg)
  File "/root/.local/lib/python3.8/site-packages/opcua/common/utils.py", line 118, in write
    self.socket.sendall(data)
OSError: [Errno 9] Bad file descriptor
INFO:opcua.client.ua_client.Socket:Cancelling request 511
INFO:opcua.client.ua_client.Socket:Cancelling request 512
INFO:opcua.client.ua_client:call_publish_callback
INFO:opcua.client.ua_client.Socket:Cancelling request 513
INFO:opcua.client.ua_client.Socket:Cancelling request 514
INFO:opcua.client.ua_client:call_publish_callback
INFO:opcua.client.ua_client.Socket:Thread ended
INFO:opcua.client.ua_client.Socket:Done closing socket: Receiving thread terminated, socket disconnected

Does that give you any ideas?
Many thanks, appreciate any help,
Mark

**Reply by AndreasHeine** [reply]
o_o' -> OSError

-> ERROR:opcua.client.ua_client:**Exception while calling user callback**: %s

**Reply by m4rkisgrov3** [reply]
So, is this a server side issue or client side?
My client is calling a callback? The handler is this? But it cannot call it? Is that correct?
Thanks.

**Reply by AndreasHeine** [reply]
the subhandler callback is called for every datachangenotification from server!
dont know whats inside your callback...

just for testing change the Subhandler:

```python
class SubHandler(object):

    """
    Subscription Handler. To receive events from server for a subscription
    data_change and event methods are called directly from receiving thread.
    Do not do expensive, slow or network operation there. Create another 
    thread if you need to do such a thing
    """

    def datachange_notification(self, node, val, data):
        print("Python: New data change event", node, val)
```

**Reply by m4rkisgrov3** [reply]
Yeap, I'll try this. Good point.
I do have quite a bit going on in the callback... I create a new thread and call a REST API endpoint within it, perhaps I'm doing too much in here?
Appreciate the help here.

**Reply by swamper123** [reply]
>     Do not do expensive, slow or network operation there. Create another 
>     thread if you need to do such a thing

With "to much" it is meant, that you shouldn't run long going processes (someone once wanted to make imageprocessing, which is slowing down the server really fast). 

>  OSError: [Errno 9] Bad file descriptor

Sounds more like another problem with your socket ... are you handling that correctly?

**Reply by m4rkisgrov3** [reply]
Thanks everyone for help here.
I've suppressed the problem and narrowed down the issue so thought it was worth noting this in the forum.

So when I'm subscribing to lots of nodes (circa 500), when I start my OPC client I get an initial surge of data change notifications. This would mean I get 500 calls to my callback at pretty much the same time. This isn't good and causes it to break with the errors above. So, to suppress this issue, I've put in a 10 seconds delay into my subscribe handler meaning that it doesn't process any nodes during this initial surge (or in the initial 10 seconds of starting the OPC client). Not ideal and it feels like a fudge, but it means I don't get that massive ramp up in activity to deal with. After the delay, it ticks along nicely and processes everything that comes in.
I'd prefer a more robust fix of which I'm still looking to explore...

@oroulet are these issues you've encountered before?

I appreciate everyone's help on this. :)
